hazelcast-client/src/main/java/com/hazelcast/client/connection/nio/ClientConnectionManagerImpl.java

```diff
@@ -192,66 +216,90 @@ protected void shutdownSelectors() {
+        BlockingCallback blockingCallback = new BlockingCallback();
+        Connection connection = getOrTriggerConnectInternal(address, asOwner, blockingCallback);
+        if (connection != null) {
+            return connection;
+        }
+        try {
+            return blockingCallback.get();
+        } catch (Throwable e) {
+            throw ExceptionUtil.rethrow(e);
+        }
+    }
+
+    private class BlockingCallback implements AuthenticationCallback {
+
+        private final CountDownLatch countDownLatch = new CountDownLatch(1);
+        private Connection connection;
+        private Throwable throwable;
 
-        if (remoteAddress == null) {
-            throw new IOException("Address is required!");
+        @Override
+        public void onSuccess(Connection connection) {
+            this.connection = connection;
+            countDownLatch.countDown();
         }
 
-        ClientConnection connection = connections.get(target);
-        Object lock = getLock(target);
+        @Override
+        public void onFailure(Throwable throwable) {
+            this.throwable = throwable;
+            countDownLatch.countDown();
+        }
 
-        if (connection == null) {
-            synchronized (lock) {
-                connection = connections.get(target);
-                if (connection == null) {
-                    connection = initializeConnection(remoteAddress, authenticator);
-                }
+        Connection get() throws Throwable {
+            countDownLatch.await();
+            if (connection != null) {
+                return connection;
             }
+            throw throwable;
         }
-        return connection;
     }
 
-    private ClientConnection initializeConnection(Address address, Authenticator authenticator) throws IOException {
-        ClientConnection connection = createSocketConnection(address);
-        authenticate(authenticator, connection);
-        connections.put(connection.getRemoteEndpoint(), connection);
-        fireConnectionAddedEvent(connection);
-        return connection;
+    interface AuthenticationCallback {
+        void onSuccess(Connection connection);
+
+        void onFailure(Throwable exception);
+    }
+
+    @Override
+    public Connection getOrTriggerConnect(Address target, boolean asOwner) {
+        return getOrTriggerConnectInternal(target, asOwner, dummyAuthCallback);
     }
 
-    public ClientConnection getOrTriggerConnect(Address target, Authenticator authenticator) throws IOException {
-        Address remoteAddress = addressTranslator.translate(target);
+    private Connection getOrTriggerConnectInternal(Address target, boolean asOwner,
+                                                   AuthenticationCallback callback) {
+        target = addressTranslator.translate(target);
 
-        if (remoteAddress == null) {
-            throw new IOException("Address is required!");
+        if (target == null) {
+            return null;
```

Comment:
Shouldn't you throw an exception here, instead of returning null? I think passing `target` as null or getting null from `addressTranslator` is an exceptional case.